### ISSUE TITLE
what happens if a cycle is not strictly monotonic

### ISSUE THREAD

**Issue by elena-pascal** [BenSmithGreyGroup/navani#15]
Hello!

I am wrong to understand that `navani` assumes a cycle is strictly monotonic because of this:
```
    df = pd.DataFrame({'Capacity': capacity, 'Voltage':voltage})
    unique_v = df.astype(float).groupby('Voltage').mean().index
    unique_v_cap = df.astype(float).groupby('Voltage').mean()['Capacity']
```

Otherwise, if there are turning points, the data gets a bit scrambled. Or is it relying on `groupby` doing a lot of heavy lifting to still find the correct peaks despite the scrambling?

Thanks

**Comment by elena-pascal** [comment]
Hi @be-smith, thanks for confirming my guess. 
I am not a battery chemistry person, but from what Alice and Jiho tell me, turning points in Voltage are not uncommon per cycle. Not only that, but the data I'm dealing with can have big turning points (as below) as well as small localised turning points. Maybe document these limitations, especially because your pipeline will not break or complain if this assumption is not satisfied. Maybe do both. At least Alice said she used the software before and didn't realise it assumes no turning points. 

Here is an example where using `groupby` is really scrambling things up

![this when not](https://github.com/user-attachments/assets/e3bbffda-ca42-4d1f-bff8-23542e1af367)

I won't be able to use Navani for my purposes, since if I am looking for an automated general approach where I don't knew where the turning points are when I start. But good luck with the repo it looks good 😃 .

**Comment by be-smith** [comment]
Yeah this is a limitation with how it currently works, I will note it in the documentation. For your case you should be able to find turning points in the data that large quite easily and simply split the data before and after those turning points and apply the dQ/dV function to each section. You could also just filter out the first 0.3 mAh of the data, I doubt there is much in the dQ/dV there that is relevant/interpretable, and with the turning points in the data no matter the approach you will have dQ/dV as undefined due to dividing by zero.

I used groupby to ensure there is only a single capacity value for each voltage value - so that a spline can be fitted to the data for differentiation.

**Comment by elena-pascal** [comment]
Do you mainly target the digital noise with `groupby`? Because it is great for that.  Your example data seems pretty clean (not noisy).

I do unfortunately have turning points that are on a scale comparable with the digital noise, that is my current pickle.
